cp1616/include/cp1616/cp1616_io_controller.h

```diff
@@ -0,0 +1,278 @@
+  /**
+   * \brief Adds output module to Controller data structure
+   *
+   * \param output_size output data length according to STEP7 setup
+   * \param output_start_address Q address according to STEP7 setup
+   *
+   * \return 0 if input module added successfully
+   */
+
+          int addOutputModule(unsigned int output_size, unsigned int output_start_address);
+
+  /**
+   * \brief Reads recieved data from IO Base layer
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicInputData();
+
+  /**
+   * \brief Writes data to IO Base library layer for transmission
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicOutputData();
+
+  /**
+   * \brief Changes IO Controller mode
+   *
+   * \param requested_mode OFFLINE/OPERATE/CLEAR
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int changeAndWaitForPnioMode(PNIO_MODE_TYPE requested_mode);
+
+  /**
+   * \brief Sets IO Controller OutData
+   *
+   * \param module output module index according to addOutputModule order
+   * \param data_index index of out_data_
+   * \param value requested value
+   */
+          void setOutData(unsigned int module, unsigned int data_index, PNIO_UINT8 value);
+
+  /**
+   * \brief Sets IO Controller cp_ready_ variable
+   *
+   * \param cp_ready_value
+   */
+          void setCpReady(int cp_ready_value);
+
+  /**
+   * \brief Returns IO Controller cp_ready_ state (used in callbackForAlarmIndication)
+   *
+   * \return cp_ready_ flag
+   */
+          int getCpReady();
+
+  /**
+   * \brief Sets IO Controller cp_current_mode_ flag (used in callbackForModeChangeIndication)
+   *
+   * \param mode - current mode
+   */
+          void setCpCurrentModeFlag(PNIO_MODE_TYPE mode);
+
+  /**
+   * \brief Returns IO Controller cp_current_mode_ flag
+   *
+   * \return cp_current_mode_
+   */
+          PNIO_MODE_TYPE getCpCurrentModeFlag();
+
+  /**
+   * \brief Sets sem_mode_change_ flag (used in callbackForModeChangeIndication)
+   *
+   * \param value
+   */
+          void setSemModChange(int mod_change);
+
+  /** Debugging functions  */
+  void printOutputData(unsigned int module);
+  void printInputData(unsigned int module);
+
+private:
+
+  /**
+   * \brief Constructor
+   */
+         Cp1616IOController();
+
+  /**
+   * \brief Static instance pointer
+   */
+         static Cp1616IOController *controller_instance_;
+
+  /**
+   * \brief CP ID according to TIA portal setup (default = 1)
+   */
+          PNIO_UINT32 cp_id_;
+
+  /**
+   * brief CP Handle obtained by PNIO_controller_open function
+   */
+          PNIO_UINT32 cp_handle_;
+
+  /**
+   * \brief CP ready communication flag
+   */
+          int cp_ready_;
+
+  /**
+   * \brief CP current mode (see PGH_IO-Base_76.pdf )
+   */
+          /*volatile*/ PNIO_MODE_TYPE cp_current_mode_;
```

Comment:
I have tested controller mode without `volatile` and noticed no problems so far, I will return it back to original definition